assume/markets/grid_utils.py

```diff
@@ -0,0 +1,131 @@
+# SPDX-FileCopyrightText: ASSUME Developers
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+import numpy as np
+import pandas as pd
+import pypsa
+
+
+def connect_grid_data(grid_dict: dict[str, pd.DataFrame]):
+    buses = grid_dict["buses"]
+
+
+def add_redispatch_generators(
+    network: pypsa.Network,
+    generators: pd.DataFrame,
+    backup_marginal_cost: float = 1e5,
+) -> None:
+    p_set = pd.DataFrame(
+        np.zeros((len(network.snapshots), len(generators.index))),
+        index=network.snapshots,
+        columns=generators.index,
+    )
+
+    # add upward redispatch generators
+    network.madd(
+        "Generator",
+        names=generators.index,
+        suffix="_up",
+        bus=generators["node"],  # bus to which the generator is connected to
+        p_nom=generators["max_power"],  # Nominal capacity of the powerplant/generator
+        p_min_pu=p_set,
+        p_max_pu=p_set + 1,
+        marginal_cost=p_set,
+    )
+
+    # add downward redispatch generators
+    network.madd(
+        "Generator",
+        names=generators.index,
+        suffix="_down",
+        bus=generators["node"],  # bus to which the generator is connected to
+        p_nom=generators["max_power"],  # Nominal capacity of the powerplant/generator
+        p_min_pu=p_set,
+        p_max_pu=p_set + 1,
+        marginal_cost=p_set,
+        sign=-1,
+    )
+
+    # add upward and downward backup generators at each node
+    network.madd(
+        "Generator",
+        names=network.buses.index,
+        suffix="_backup_up",
+        bus=network.buses.index,  # bus to which the generator is connected to
+        p_nom=10e4,
+        marginal_cost=backup_marginal_cost,
+    )
+
+    network.madd(
+        "Generator",
+        names=network.buses.index,
+        suffix="_backup_down",
+        bus=network.buses.index,  # bus to which the generator is connected to
+        p_nom=10e4,
+        marginal_cost=backup_marginal_cost,
+        sign=-1,
+    )
+
+
+def read_pypsa_grid(
+    network: pypsa.Network,
+    grid_dict: dict[str, pd.DataFrame],
+    backup_marginal_cost: float = 1e5,
+):
+    def add_buses(network: pypsa.Network, buses: pd.DataFrame) -> None:
+        network.madd(
+            "Bus",
+            names=buses.index,
+            **buses,
+        )
+
+    def add_lines(network: pypsa.Network, lines: pd.DataFrame) -> None:
+        network.madd(
+            "Line",
+            names=lines.index,
+            **lines,
+        )
+
+    def add_generators(
```

Comment:
this function is also specific for redispatch, since generators are loaded as Load. This won't be the case for normal networks

Reply:
I hope I fixed this as expected.